{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a88e1737-aea5-4332-86fc-541f4ff7a821",
   "metadata": {},
   "source": [
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fc3c49f0-2dcf-49e3-86d6-a1bb79055a64",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/in68/miniconda3/envs/cell2loc_env/lib/python3.9/site-packages/umap/distances.py:1063: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n",
      "  @numba.jit()\n",
      "/home/in68/miniconda3/envs/cell2loc_env/lib/python3.9/site-packages/umap/distances.py:1071: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n",
      "  @numba.jit()\n",
      "/home/in68/miniconda3/envs/cell2loc_env/lib/python3.9/site-packages/umap/distances.py:1086: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n",
      "  @numba.jit()\n",
      "/home/in68/.local/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/home/in68/miniconda3/envs/cell2loc_env/lib/python3.9/site-packages/umap/umap_.py:660: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n",
      "  @numba.jit()\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy\n",
    "import os\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import gc\n",
    "from scipy.stats import pearsonr\n",
    "import os\n",
    "import scanpy as sc\n",
    "import muon as mu\n",
    "from scipy.sparse import issparse, csr_matrix\n",
    "from typing import Optional\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import matplotlib.font_manager as fm\n",
    "\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from scipy.cluster.hierarchy import linkage, leaves_list\n",
    "from scipy.sparse import issparse\n",
    "from scipy.spatial import distance_matrix\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "gc.enable()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f2f75ce1-bc6e-4e11-94b5-d7ff120e0a62",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utility_scripts import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c342230f-c86c-4335-81b5-c3831b9d76b7",
   "metadata": {},
   "source": [
    "## Databases: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4e997616-fd3f-4538-b5db-ea220b6c3d2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using 56973 ensembl IDs, from 56825 genes, and 38 biotypes...\n",
      "\n",
      "                 GENEID     GeneSymbol            Chromosome  \\\n",
      "0  ENSMUSG00000102693.2  4933401J01Rik  chr1:3143476-3144545   \n",
      "1  ENSMUSG00000064842.3        Gm26206  chr1:3172239-3172348   \n",
      "2  ENSMUSG00000051951.6           Xkr4  chr1:3276124-3741721   \n",
      "3  ENSMUSG00000102851.2        Gm18956  chr1:3322980-3323459   \n",
      "4  ENSMUSG00000103377.2        Gm37180  chr1:3435954-3438772   \n",
      "\n",
      "                Biotype Strand              GeneID   chr  \n",
      "0                   TEC      +  ENSMUSG00000102693  chr1  \n",
      "1                 snRNA      +  ENSMUSG00000064842  chr1  \n",
      "2        protein_coding      -  ENSMUSG00000051951  chr1  \n",
      "3  processed_pseudogene      +  ENSMUSG00000102851  chr1  \n",
      "4                   TEC      -  ENSMUSG00000103377  chr1  \n",
      "    tax_id superkingdom          phylum                class  \\\n",
      "0  taxid_1          NaN             NaN                  NaN   \n",
      "1  taxid_2     Bacteria             NaN                  NaN   \n",
      "2  taxid_6     Bacteria  Pseudomonadota  Alphaproteobacteria   \n",
      "3  taxid_7     Bacteria  Pseudomonadota  Alphaproteobacteria   \n",
      "4  taxid_9     Bacteria  Pseudomonadota  Gammaproteobacteria   \n",
      "\n",
      "              order             family         genus  \\\n",
      "0               NaN                NaN           NaN   \n",
      "1               NaN                NaN           NaN   \n",
      "2  Hyphomicrobiales  Xanthobacteraceae  Azorhizobium   \n",
      "3  Hyphomicrobiales  Xanthobacteraceae  Azorhizobium   \n",
      "4  Enterobacterales        Erwiniaceae      Buchnera   \n",
      "\n",
      "                    species biotype clade  ... subphylum subsection  \\\n",
      "0                       NaN     NaN   NaN  ...       NaN        NaN   \n",
      "1                       NaN     NaN   NaN  ...       NaN        NaN   \n",
      "2                       NaN     NaN   NaN  ...       NaN        NaN   \n",
      "3  Azorhizobium caulinodans     NaN   NaN  ...       NaN        NaN   \n",
      "4       Buchnera aphidicola     NaN   NaN  ...       NaN        NaN   \n",
      "\n",
      "  subspecies subtribe superclass superfamily superorder superphylum tribe  \\\n",
      "0        NaN      NaN        NaN         NaN        NaN         NaN   NaN   \n",
      "1        NaN      NaN        NaN         NaN        NaN         NaN   NaN   \n",
      "2        NaN      NaN        NaN         NaN        NaN         NaN   NaN   \n",
      "3        NaN      NaN        NaN         NaN        NaN         NaN   NaN   \n",
      "4        NaN      NaN        NaN         NaN        NaN         NaN   NaN   \n",
      "\n",
      "  varietas  \n",
      "0      NaN  \n",
      "1      NaN  \n",
      "2      NaN  \n",
      "3      NaN  \n",
      "4      NaN  \n",
      "\n",
      "[5 rows x 70 columns]\n"
     ]
    }
   ],
   "source": [
    "#Load and process the gene annotation file to use metadata for analysis e.g. Biotype % calculations\n",
    "# Read the .tsv file\n",
    "gtf_info = pd.read_csv(\"Annotation_files/GRCm39_GENCODEM32_ReoT1L_gene_info_gtf.tsv\", sep=\"\\t\")\n",
    "gtf_info = gtf_info.drop_duplicates(subset=['GENEID'], keep='first')\n",
    "\n",
    "# Remove period/decimal suffixes from ensembl IDs\n",
    "gtf_info['GeneID'] = gtf_info['GENEID'].str.split('.').str[0]\n",
    "\n",
    "# Extract chromosome value before the colon\n",
    "gtf_info['chr'] = gtf_info['Chromosome'].str.split(':').str[0]\n",
    "\n",
    "# Add '.1' to the Reovirus gene names to match the gtf file...\n",
    "gtf_info[\"GENEID\"] = [f\"{GENE}.1\" if 'T1LReovirus' in GENE else GENE for GENE in gtf_info['GENEID']]\n",
    "\n",
    "print(f\"Using {len(gtf_info['GENEID'].unique())} ensembl IDs, from {gtf_info['GeneSymbol'].nunique()} genes, and {gtf_info['Biotype'].nunique()} biotypes...\\n\")\n",
    "\n",
    "# Display the first few rows\n",
    "print(gtf_info.head())\n",
    "\n",
    "\n",
    "\n",
    "#Add NCBI taxonomic ID file as dataframe to build taxonomic-level specific adata objects \n",
    "taxid = pd.read_csv('Annotation_files/ncbi_lineages_2024-01-19.csv')\n",
    "Taxonomic_df = taxid.copy()\n",
    "def add_prefix(x):\n",
    "    return f'taxid_{x}'\n",
    "\n",
    "# Applying function to 'tax_id' column\n",
    "Taxonomic_df['tax_id'] = Taxonomic_df['tax_id'].apply(add_prefix)\n",
    "\n",
    "# Display the updated DataFrame\n",
    "print(Taxonomic_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2804fb7-a463-4ff6-9816-d3b269b1cde0",
   "metadata": {},
   "source": [
    "### Create objects "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7830707f-2eaa-4996-8da0-3fee89c256bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxid_to_delete = ['taxid_1', 'taxid_10090', 'taxid_119060', 'taxid_1211813', 'taxid_1224',\n",
    "       'taxid_1236', 'taxid_1239', 'taxid_1268', 'taxid_1269', 'taxid_1279',\n",
    "       'taxid_1280', 'taxid_1301', 'taxid_131567', 'taxid_1351', 'taxid_1385',\n",
    "       'taxid_1396', 'taxid_1491', 'taxid_153135', 'taxid_1624',\n",
    "       'taxid_165779', 'taxid_1716', 'taxid_1745854', 'taxid_1747',\n",
    "       'taxid_1760', 'taxid_1783272', 'taxid_1789224', 'taxid_1853699',\n",
    "       'taxid_186817', 'taxid_1883', 'taxid_190721', 'taxid_1912216',\n",
    "       'taxid_191302', 'taxid_1916956', 'taxid_2', 'taxid_201174',\n",
    "       'taxid_2049881', 'taxid_2184519', 'taxid_2306583', 'taxid_2618356',\n",
    "       'taxid_2654547', 'taxid_2712222', 'taxid_2722752', 'taxid_2726118',\n",
    "       'taxid_2782662', 'taxid_28211', 'taxid_28216', 'taxid_286',\n",
    "       'taxid_2862880', 'taxid_2892442', 'taxid_2986923', 'taxid_2986930',\n",
    "       'taxid_314146', 'taxid_31957', 'taxid_31989', 'taxid_32207',\n",
    "       'taxid_33011', 'taxid_33958', 'taxid_358', 'taxid_36845', 'taxid_40324',\n",
    "       'taxid_42906', 'taxid_43675', 'taxid_44283', 'taxid_446469',\n",
    "       'taxid_465541', 'taxid_469', 'taxid_471', 'taxid_48736', 'taxid_524884',\n",
    "       'taxid_525919', 'taxid_543', 'taxid_55601', 'taxid_564137', 'taxid_570',\n",
    "       'taxid_573', 'taxid_59732', 'taxid_80840', 'taxid_82541', 'taxid_85006',\n",
    "       'taxid_881', 'taxid_91061', 'taxid_91347', 'taxid_93386', 'taxid_9606',\n",
    "       'taxid_976']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c1d180f5-0923-4d01-83c1-2653735c1e2d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "No gene_version1 genes found...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "Added spatial coordinates for 4991 cells.\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "No gene_version1 genes found...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "Added spatial coordinates for 4986 cells.\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "No gene_version1 genes found...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "No gene_version1 genes found...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "Added spatial coordinates for 4966 cells.\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "No gene_version1 genes found...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "Added spatial coordinates for 4975 cells.\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "No gene_version1 genes found...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "Added spatial coordinates for 4973 cells.\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "No gene_version1 genes found...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "Added spatial coordinates for 4810 cells.\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "No gene_version1 genes found...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Processed samples: ['STRS_A', 'STRS_B', 'STRS_C', 'STRS_D', 'Visium_A', 'Visium_B', 'Visium_C', 'Visium_D']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "10075"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the taxonomic identifiers for bacteria from a DataFrame\n",
    "bacteria_taxids = Taxonomic_df[Taxonomic_df['superkingdom'] == 'Bacteria']['tax_id']\n",
    "viruses_taxids = Taxonomic_df[Taxonomic_df['superkingdom'] == 'Viruses']['tax_id']\n",
    "archaea_taxids = Taxonomic_df[Taxonomic_df['superkingdom'] == 'Archaea']['tax_id']\n",
    "\n",
    "# Define the base directory where the host and microbial data are stored\n",
    "base_dir = \"/workdir/in68/microSTRS/results_out_Mar24/Mar24_KrakenDB/\"\n",
    "spatial_coord_path = \"/workdir/lt425/202403_visium/visium-v1_coordinates.txt\"\n",
    "# List all the sample identifiers\n",
    "samples = ['STRS_A',\n",
    "           'STRS_B', 'STRS_C', 'STRS_D','Visium_A', 'Visium_B', 'Visium_C', 'Visium_D'\n",
    "          ]\n",
    "# Dictionary to store the combined data objects (MuData) for each sample\n",
    "mudata_dict = {}\n",
    "\n",
    "# Process data for each sample\n",
    "for sample in samples:\n",
    "    try:\n",
    "        # Assemble the directory path for the host data for the current sample\n",
    "        host_data_dir = os.path.join(base_dir, f\"Sample_{sample}_solo/Solo.out/GeneFull/raw/\")\n",
    "        spliced_data_dir = os.path.join(base_dir, f\"Sample_{sample}_solo/Solo.out/Velocyto/raw/spliced.mtx.gz\")\n",
    "        unspliced_data_dir = os.path.join(base_dir, f\"Sample_{sample}_solo/Solo.out/Velocyto/raw/unspliced.mtx.gz\")\n",
    "        \n",
    "        # Check if the host data directory exists; skip the sample if it doesn't\n",
    "        if not os.path.exists(host_data_dir):\n",
    "            print(f\"Host data directory does not exist: {host_data_dir}\")\n",
    "            continue\n",
    "\n",
    "        # Read the host gene expression data using the 10x Genomics format\n",
    "        adata_host = sc.read_10x_mtx(host_data_dir, var_names='gene_symbols')\n",
    "        adata_host = normalize_spatial_coordinates(adata_host, spatial_coord_path)\n",
    "        adata_host.raw = adata_host.copy()\n",
    "        # Add biotypes percentage and total counts to the host data\n",
    "        add_biotypes_pct(adata_host, gtf_info)\n",
    "        adata_host.obs['Total_counts'] = adata_host.X.sum(axis=1)\n",
    "        \n",
    "        # Normalize total gene expression and log transform for host data\n",
    "        sc.pp.normalize_total(adata_host, target_sum=1e4)\n",
    "        sc.pp.log1p(adata_host)\n",
    "\n",
    "        # Read spliced and unspliced data\n",
    "        adata_spliced = sc.read_mtx(spliced_data_dir)\n",
    "        adata_unspliced = sc.read_mtx(unspliced_data_dir)\n",
    "        adata_spliced = adata_spliced.T\n",
    "        adata_unspliced = adata_unspliced.T\n",
    "        adata_spliced.obs = adata_host.obs.copy()\n",
    "        adata_unspliced.obs = adata_host.obs.copy()\n",
    "        adata_spliced.var = adata_host.var.copy()\n",
    "        adata_unspliced.var = adata_host.var.copy()\n",
    "        adata_host.obs['Total_counts_spliced'] = adata_spliced.X.sum(axis=1)\n",
    "        adata_host.obs['Total_counts_unspliced'] = adata_unspliced.X.sum(axis=1)\n",
    "        adata_host.obs['Unspliced.pct'] = adata_host.obs['Total_counts_unspliced']/(adata_host.obs['Total_counts_spliced']+adata_host.obs['Total_counts_unspliced'])\n",
    "\n",
    "        # Construct the path for microbial data and verify it\n",
    "        microbial_data_path = os.path.join(base_dir, f\"Sample_{sample}_solo/{sample}_microbe.h5ad\")\n",
    "        \n",
    "        # Check if the microbial data file exists; skip the sample if it doesn't\n",
    "        if not os.path.exists(microbial_data_path):\n",
    "            print(f\"Microbial data file does not exist: {microbial_data_path}\")\n",
    "            continue\n",
    "\n",
    "        # Read the microbial data\n",
    "        adata_unmapped = sc.read_h5ad(microbial_data_path)\n",
    "        adata_unmapped = normalize_spatial_coordinates(adata_unmapped, spatial_coord_path)\n",
    "\n",
    "        # Store the raw microbial data\n",
    "        adata_unmapped.raw = adata_unmapped.copy()\n",
    "        taxonomy_tmp = Taxonomic_df.copy()\n",
    "        adata_unmapped.obs['Total_counts_unfiltered'] = adata_unmapped.X.sum(axis=1)\n",
    "\n",
    "        # Filter the microbial data to include only bacteria\n",
    "        mask = adata_unmapped.var.index.isin(bacteria_taxids)\n",
    "        adata_bacteria = adata_unmapped[:, mask].copy()\n",
    "        adata_bacteria.obs['Total_counts_unfiltered'] = adata_bacteria.X.sum(axis=1)\n",
    "\n",
    "        # Delete the taxids that are found in controls\n",
    "        mask2 = adata_bacteria.var.index.isin(taxid_to_delete)\n",
    "        adata_bacteria = adata_bacteria[:, ~mask2].copy()\n",
    "\n",
    "\n",
    "        \n",
    "        # Filter the microbial data to include only virus\n",
    "        mask = adata_unmapped.var.index.isin(viruses_taxids)\n",
    "        adata_virus = adata_unmapped[:, mask].copy()\n",
    "        adata_virus.obs['Total_counts_unfiltered'] = adata_virus.X.sum(axis=1)\n",
    "\n",
    "        # Delete the taxids that are found in controls from virus\n",
    "        mask2 = adata_virus.var.index.isin(taxid_to_delete)\n",
    "        adata_virus = adata_virus[:, ~mask2].copy()\n",
    "        adata_virus.obs['Total_counts'] = adata_virus.X.sum(axis=1)\n",
    "\n",
    "\n",
    "        \n",
    "        # Filter the microbial data to include only ARCHAEA\n",
    "        mask = adata_unmapped.var.index.isin(archaea_taxids)\n",
    "        adata_archaea = adata_unmapped[:, mask].copy()\n",
    "        adata_archaea.obs['Total_counts_unfiltered'] = adata_archaea.X.sum(axis=1)\n",
    "\n",
    "        # Delete the taxids that are found in controls from virus\n",
    "        mask2 = adata_archaea.var.index.isin(taxid_to_delete)\n",
    "        adata_archaea = adata_archaea[:, ~mask2].copy()\n",
    "        adata_archaea.obs['Total_counts'] = adata_archaea.X.sum(axis=1)\n",
    "\n",
    "\n",
    "        \n",
    "        adata_host.obs['Total_molecules_detected'] = adata_host.obs['Total_counts'] + adata_unmapped.obs['Total_counts_unfiltered']\n",
    "        adatas = {'host': adata_host, 'bacteria': adata_bacteria.copy(),\n",
    "                  'viruses': adata_virus.copy(),'archaea': adata_archaea.copy(), \n",
    "                  'unmapped_all': adata_unmapped.copy(), 'spliced': adata_spliced, 'unspliced': adata_unspliced}\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "        # Process each taxonomic level bacteria\n",
    "        for tax_level in ['species', 'genus', 'family', 'phylum']:\n",
    "            # adata_bacteria.uns['taxonomy'][tax_level] = adata_ba el].fillna('unclassified')\n",
    "            tax_id_to_name = dict(zip(Taxonomic_df['tax_id'], Taxonomic_df[tax_level]))\n",
    "            adata_bacteria.var[tax_level] = adata_bacteria.var.index.map(tax_id_to_name).fillna('unclassified')\n",
    "            adata_df = pd.DataFrame(adata_bacteria.X.toarray(), columns=adata_bacteria.var[tax_level])\n",
    "            aggregated_data = adata_df.groupby(adata_df.columns, axis=1).sum()\n",
    "\n",
    "            new_vars = pd.DataFrame(index=aggregated_data.columns)\n",
    "            X_mat = csr_matrix(np.round(aggregated_data.values).astype(np.float32))\n",
    "\n",
    "            new_adata_bacteria = sc.AnnData(X=X_mat, var=new_vars, obs=adata_bacteria.obs)\n",
    "\n",
    "            # Add the raw component\n",
    "            new_adata_bacteria.raw = new_adata_bacteria.copy()\n",
    "\n",
    "            new_adata_bacteria.obs['Total_counts'] = new_adata_bacteria.X.sum(axis=1)\n",
    "            unclassified_mask = new_adata_bacteria.var.index == 'unclassified'\n",
    "            new_adata_bacteria.obs['Total_counts_unclassified'] = new_adata_bacteria[:, unclassified_mask].X.sum(axis=1)\n",
    "            new_adata_bacteria.obs['Total_counts_classified'] = new_adata_bacteria.obs['Total_counts'] - new_adata_bacteria.obs['Total_counts_unclassified']\n",
    "            # Filtering step after calculating Total_counts\n",
    "            # total_counts = new_adata_bacteria.obs['Total_counts'].sum()\n",
    "            # taxon_percentages = new_adata_bacteria.X.sum(axis=0) / total_counts * 100\n",
    "\n",
    "            # # Identify taxa that are <= 0.01% and remove them\n",
    "            # low_abundance_mask = taxon_percentages <= 0.01\n",
    "            # new_adata_bacteria = new_adata_bacteria[:, ~low_abundance_mask].copy()\n",
    "            # # Remove 'unclassified' variable if it exists\n",
    "            # if 'unclassified' in new_adata_bacteria.var_names:\n",
    "            #     unclassified_mask = new_adata_bacteria.var_names == 'unclassified'\n",
    "            #     new_adata_bacteria = new_adata_bacteria[:, ~unclassified_mask].copy()\n",
    "\n",
    "\n",
    "            \n",
    "            # Normalize counts and log transform\n",
    "            total_microbial_counts = new_adata_bacteria.X.sum()\n",
    "            new_adata_bacteria.X = new_adata_bacteria.X / total_microbial_counts * 10000\n",
    "            new_adata_bacteria = normalize_spatial_coordinates(new_adata_bacteria, spatial_coord_path)\n",
    "            spot_counts = new_adata_bacteria.X > 0\n",
    "            unique_genes_per_spot = spot_counts.sum(axis=1)\n",
    "            new_adata_bacteria.obs['unique_taxa_per_spot'] = unique_genes_per_spot\n",
    "            sc.pp.log1p(new_adata_bacteria)\n",
    "\n",
    "            adatas[tax_level] = new_adata_bacteria\n",
    "            # Delete intermediate variables to free up memory\n",
    "            del adata_df, aggregated_data, new_vars, X_mat, new_adata_bacteria, tax_id_to_name\n",
    "       \n",
    "\n",
    "        # Process each taxonomic level for virus\n",
    "        for tax_level in ['species', 'genus', 'family', 'phylum']:\n",
    "            # adata_virus.uns['taxonomy'][tax_level] = adata_virus.uns['taxonomy'][tax_level].fillna('unclassified')\n",
    "            tax_id_to_name = dict(zip(Taxonomic_df['tax_id'], Taxonomic_df[tax_level]))\n",
    "            adata_virus.var[tax_level] = adata_virus.var.index.map(tax_id_to_name).fillna('unclassified')\n",
    "            adata_df = pd.DataFrame(adata_virus.X.toarray(), columns=adata_virus.var[tax_level])\n",
    "            aggregated_data = adata_df.groupby(adata_df.columns, axis=1).sum()\n",
    "\n",
    "            new_vars = pd.DataFrame(index=aggregated_data.columns)\n",
    "            X_mat = csr_matrix(np.round(aggregated_data.values).astype(np.float32))\n",
    "\n",
    "            new_adata_virus = sc.AnnData(X=X_mat, var=new_vars, obs=adata_virus.obs)\n",
    "\n",
    "            # Add the raw component\n",
    "            new_adata_virus.raw = new_adata_virus.copy()\n",
    "\n",
    "            new_adata_virus.obs['Total_counts'] = new_adata_virus.X.sum(axis=1)\n",
    "            unclassified_mask = new_adata_virus.var.index == 'unclassified'\n",
    "            new_adata_virus.obs['Total_counts_unclassified'] = new_adata_virus[:, unclassified_mask].X.sum(axis=1)\n",
    "            new_adata_virus.obs['Total_counts_classified'] = new_adata_virus.obs['Total_counts'] - new_adata_virus.obs['Total_counts_unclassified']\n",
    "\n",
    "            # Normalize counts and log transform\n",
    "            total_microbial_counts = new_adata_virus.X.sum()\n",
    "            new_adata_virus.X = new_adata_virus.X / total_microbial_counts * 10000\n",
    "            new_adata_virus = normalize_spatial_coordinates(new_adata_virus, spatial_coord_path)\n",
    "            spot_counts = new_adata_virus.X > 0\n",
    "            unique_genes_per_spot = spot_counts.sum(axis=1)\n",
    "            new_adata_virus.obs['unique_taxa_per_spot'] = unique_genes_per_spot\n",
    "            sc.pp.log1p(new_adata_virus)\n",
    "            tax_level_to_use = f\"{tax_level}_virus\"\n",
    "            adatas[tax_level_to_use] = new_adata_virus\n",
    "            # Delete intermediate variables to free up memory\n",
    "            del adata_df, aggregated_data, new_vars, X_mat, new_adata_virus, tax_id_to_name\n",
    "            \n",
    "        # Process each taxonomic level for archaea\n",
    "        for tax_level in ['species', 'genus', 'family', 'phylum']:\n",
    "            # adata_archaea.uns['taxonomy'][tax_level] = adata_archaea.uns['taxonomy'][tax_level].fillna('unclassified')\n",
    "            tax_id_to_name = dict(zip(Taxonomic_df['tax_id'], Taxonomic_df[tax_level]))\n",
    "            adata_archaea.var[tax_level] = adata_archaea.var.index.map(tax_id_to_name).fillna('unclassified')\n",
    "            adata_df = pd.DataFrame(adata_archaea.X.toarray(), columns=adata_archaea.var[tax_level])\n",
    "            aggregated_data = adata_df.groupby(adata_df.columns, axis=1).sum()\n",
    "\n",
    "            new_vars = pd.DataFrame(index=aggregated_data.columns)\n",
    "            X_mat = csr_matrix(np.round(aggregated_data.values).astype(np.float32))\n",
    "\n",
    "            new_adata_archaea = sc.AnnData(X=X_mat, var=new_vars, obs=adata_archaea.obs)\n",
    "\n",
    "            # Add the raw component\n",
    "            new_adata_archaea.raw = new_adata_archaea.copy()\n",
    "\n",
    "            new_adata_archaea.obs['Total_counts'] = new_adata_archaea.X.sum(axis=1)\n",
    "            unclassified_mask = new_adata_archaea.var.index == 'unclassified'\n",
    "            new_adata_archaea.obs['Total_counts_unclassified'] = new_adata_archaea[:, unclassified_mask].X.sum(axis=1)\n",
    "            new_adata_archaea.obs['Total_counts_classified'] = new_adata_archaea.obs['Total_counts'] - new_adata_archaea.obs['Total_counts_unclassified']\n",
    "\n",
    "            # Normalize counts and log transform\n",
    "            total_microbial_counts = new_adata_archaea.X.sum()\n",
    "            new_adata_archaea.X = new_adata_archaea.X / total_microbial_counts * 10000\n",
    "            new_adata_archaea = normalize_spatial_coordinates(new_adata_archaea, spatial_coord_path)\n",
    "            spot_counts = new_adata_archaea.X > 0\n",
    "            unique_genes_per_spot = spot_counts.sum(axis=1)\n",
    "            new_adata_archaea.obs['unique_taxa_per_spot'] = unique_genes_per_spot\n",
    "            sc.pp.log1p(new_adata_archaea)\n",
    "            tax_level_to_use = f\"{tax_level}_archaea\"\n",
    "            adatas[tax_level_to_use] = new_adata_archaea\n",
    "\n",
    "\n",
    "\n",
    "            \n",
    "            # Delete intermediate variables to free up memory\n",
    "            del adata_df, aggregated_data, new_vars, X_mat, new_adata_archaea, tax_id_to_name\n",
    "\n",
    "        # Store the combined data object for each sample\n",
    "        mudata_dict[sample] = mu.MuData(adatas)\n",
    "\n",
    "        # Delete intermediate variables to free up memory\n",
    "        del adata_host, adata_unmapped, adata_bacteria, adatas, mask, mask2, adata_spliced, adata_unspliced\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred while processing sample {sample}: {e}\")\n",
    "\n",
    "# Output which samples have been successfully processed\n",
    "print(\"Processed samples:\", list(mudata_dict.keys()))\n",
    "gc.collect()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed98858a-3b08-4b66-90a7-fc281138bdc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define paths to JSON files and barcode file for each sample\n",
    "samples_json_files = {\n",
    "    'STRS_A': [\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-A1_Outside.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-A1_Lumen.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-A1_Tissue.json\",\n",
    "    ],\n",
    "    'STRS_B': [\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-B1_Outside.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-B1_Lumen.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-B1_Tissue.json\",\n",
    "    ],\n",
    "    'STRS_C': [\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-C1_Outside.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-C1_Lumen.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-C1_Tissue.json\",\n",
    "    ],\n",
    "    'STRS_D': [\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-D1_Outside.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-D1_Lumen.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/STRS_2024/V13A05-372-D1_Tissue.json\",\n",
    "    ],    \n",
    "    'Visium_A': [\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-A1_Outside.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-A1_Lumen.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-A1_Tissue.json\",\n",
    "    ],\n",
    "    'Visium_B': [\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-B1_Outside.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-B1_Lumen.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-B1_Tissue.json\",\n",
    "    ],\n",
    "    'Visium_C': [\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-C1_Outside.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-C1_Lumen.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-C1_Tissue.json\",\n",
    "    ],\n",
    "    'Visium_D': [\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-D1_Outside.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-D1_Lumen.json\",\n",
    "        \"/workdir/in68/micro_STRS_github/microSTRS/Analysis/Low_resolution/tissue_location_files/Visium_2024/V13A05-407-D1_Tissue.json\",\n",
    "    ]    \n",
    "}\n",
    "\n",
    "# Path to the barcode file\n",
    "barcode_file = spatial_coord_path\n",
    "\n",
    "# Example MuData objects (assuming these are pre-loaded)\n",
    "\n",
    "\n",
    "for sample, json_files in samples_json_files.items():\n",
    "    # Process the JSON files and create the DataFrame\n",
    "    barcode_presence_df = add_histology_info_visium(json_files, barcode_file)\n",
    "    \n",
    "    # Add the classification DataFrame to the `obsm` slot of each modality in the MuData object\n",
    "    mudata = mudata_dict[sample]\n",
    "    for modality in mudata.mod.keys():\n",
    "        adata = mudata[modality]\n",
    "        adata.obsm['classification'] = barcode_presence_df.loc[adata.obs_names].fillna(False)\n",
    "\n",
    "# Now each MuData object in `mudata_objects` has the barcode classification added to its modalities' `obsm` slots.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a08f25c3-5ff9-4e9d-b398-454d0dcfdcb0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample STRS_A:\n",
      "AnnData object with n_obs × n_vars = 2701 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample STRS_B:\n",
      "AnnData object with n_obs × n_vars = 1662 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample STRS_C:\n",
      "AnnData object with n_obs × n_vars = 1642 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample STRS_D:\n",
      "AnnData object with n_obs × n_vars = 699 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample Visium_A:\n",
      "AnnData object with n_obs × n_vars = 1553 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample Visium_B:\n",
      "AnnData object with n_obs × n_vars = 1725 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample Visium_C:\n",
      "AnnData object with n_obs × n_vars = 1056 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample Visium_D:\n",
      "AnnData object with n_obs × n_vars = 604 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample STRS_A:\n",
      "MuData object with n_obs × n_vars = 4992 × 178246\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4991 x 2067\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4991 x 26\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4991 x 49\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4991 x 2214\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4991 x 1624\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4991 x 862\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4991 x 340\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4991 x 37\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4991 x 26\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4991 x 20\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4991 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4991 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4991 x 41\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4991 x 33\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4991 x 21\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4991 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t2701 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample STRS_B:\n",
      "MuData object with n_obs × n_vars = 4992 × 178406\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4986 x 2137\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4986 x 16\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4986 x 37\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4986 x 2262\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4986 x 1684\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4986 x 895\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4986 x 343\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4986 x 38\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4986 x 16\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4986 x 13\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4986 x 12\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4986 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4986 x 33\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4986 x 31\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4986 x 19\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4986 x 5\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t1662 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample STRS_C:\n",
      "MuData object with n_obs × n_vars = 4992 × 188867\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4992 x 5564\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4992 x 29\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4992 x 66\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4992 x 5747\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4992 x 4255\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4992 x 1604\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4992 x 497\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4992 x 46\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4992 x 22\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4992 x 25\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4992 x 17\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4992 x 7\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4992 x 53\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4992 x 44\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4992 x 26\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4992 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t1642 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample STRS_D:\n",
      "MuData object with n_obs × n_vars = 4992 × 191543\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4966 x 6472\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4966 x 20\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4966 x 83\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4966 x 6667\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4966 x 4996\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4966 x 1694\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4966 x 502\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4966 x 46\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4966 x 19\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4966 x 20\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4966 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4966 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4966 x 61\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4966 x 49\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4966 x 26\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4966 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t699 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample Visium_A:\n",
      "MuData object with n_obs × n_vars = 4992 × 173621\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4975 x 707\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4975 x 13\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4975 x 12\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4975 x 777\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4975 x 605\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4975 x 379\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4975 x 180\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4975 x 22\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4975 x 13\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4975 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4975 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4975 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4975 x 12\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4975 x 9\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4975 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4975 x 3\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t1553 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample Visium_B:\n",
      "MuData object with n_obs × n_vars = 4992 × 174167\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4973 x 871\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4973 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4973 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4973 x 953\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4973 x 706\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4973 x 436\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4973 x 207\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4973 x 25\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4973 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4973 x 10\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4973 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4973 x 5\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4973 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4973 x 13\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4973 x 12\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4973 x 4\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t1725 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample Visium_C:\n",
      "MuData object with n_obs × n_vars = 4992 × 175366\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4810 x 1219\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4810 x 11\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4810 x 16\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4810 x 1304\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4810 x 966\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4810 x 607\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4810 x 275\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4810 x 33\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4810 x 11\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4810 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4810 x 7\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4810 x 5\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4810 x 16\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4810 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4810 x 10\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4810 x 4\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t1056 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample Visium_D:\n",
      "MuData object with n_obs × n_vars = 4992 × 176432\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4887 x 1583\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4887 x 9\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4887 x 11\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4887 x 1658\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4887 x 1164\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4887 x 728\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4887 x 323\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4887 x 41\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4887 x 9\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4887 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4887 x 5\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4887 x 5\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4887 x 11\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4887 x 10\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4887 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4887 x 2\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t604 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample STRS_A:\n",
      "MuData object with n_obs × n_vars = 4992 × 178246\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4991 x 2067\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4991 x 26\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4991 x 49\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4991 x 2214\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4991 x 1624\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4991 x 862\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4991 x 340\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4991 x 37\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4991 x 26\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4991 x 20\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4991 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4991 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4991 x 41\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4991 x 33\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4991 x 21\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4991 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t2701 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample STRS_B:\n",
      "MuData object with n_obs × n_vars = 4992 × 178406\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4986 x 2137\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4986 x 16\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4986 x 37\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4986 x 2262\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4986 x 1684\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4986 x 895\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4986 x 343\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4986 x 38\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4986 x 16\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4986 x 13\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4986 x 12\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4986 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4986 x 33\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4986 x 31\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4986 x 19\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4986 x 5\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t1662 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample STRS_C:\n",
      "MuData object with n_obs × n_vars = 4992 × 188867\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4992 x 5564\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4992 x 29\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4992 x 66\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4992 x 5747\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4992 x 4255\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4992 x 1604\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4992 x 497\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4992 x 46\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4992 x 22\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4992 x 25\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4992 x 17\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4992 x 7\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4992 x 53\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4992 x 44\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4992 x 26\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4992 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t1642 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample STRS_D:\n",
      "MuData object with n_obs × n_vars = 4992 × 191543\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4966 x 6472\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4966 x 20\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4966 x 83\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4966 x 6667\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4966 x 4996\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4966 x 1694\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4966 x 502\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4966 x 46\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4966 x 19\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4966 x 20\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4966 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4966 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4966 x 61\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4966 x 49\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4966 x 26\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4966 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t699 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample Visium_A:\n",
      "MuData object with n_obs × n_vars = 4992 × 173621\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4975 x 707\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4975 x 13\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4975 x 12\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4975 x 777\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4975 x 605\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4975 x 379\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4975 x 180\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4975 x 22\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4975 x 13\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4975 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4975 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4975 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4975 x 12\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4975 x 9\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4975 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4975 x 3\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t1553 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample Visium_B:\n",
      "MuData object with n_obs × n_vars = 4992 × 174167\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4973 x 871\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4973 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4973 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4973 x 953\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4973 x 706\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4973 x 436\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4973 x 207\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4973 x 25\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4973 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4973 x 10\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4973 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4973 x 5\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4973 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4973 x 13\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4973 x 12\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4973 x 4\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t1725 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample Visium_C:\n",
      "MuData object with n_obs × n_vars = 4992 × 175366\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4810 x 1219\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4810 x 11\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4810 x 16\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4810 x 1304\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4810 x 966\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4810 x 607\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4810 x 275\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4810 x 33\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4810 x 11\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4810 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4810 x 7\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4810 x 5\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4810 x 16\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4810 x 15\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4810 x 10\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4810 x 4\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t1056 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample Visium_D:\n",
      "MuData object with n_obs × n_vars = 4992 × 176432\n",
      "  20 modalities\n",
      "    host:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced.pct', 'Total_molecules_detected'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    bacteria:\t4887 x 1583\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    viruses:\t4887 x 9\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    archaea:\t4887 x 11\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    unmapped_all:\t4887 x 1658\n",
      "      obs:\t'Total_counts_unfiltered'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    spliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    unspliced:\t4992 x 56953\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts'\n",
      "      var:\t'gene_ids', 'feature_types'\n",
      "      obsm:\t'classification'\n",
      "    species:\t4887 x 1164\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus:\t4887 x 728\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family:\t4887 x 323\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum:\t4887 x 41\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_virus:\t4887 x 9\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_virus:\t4887 x 6\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_virus:\t4887 x 5\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_virus:\t4887 x 5\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    species_archaea:\t4887 x 11\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    genus_archaea:\t4887 x 10\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    family_archaea:\t4887 x 8\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    phylum_archaea:\t4887 x 2\n",
      "      obs:\t'Total_counts_unfiltered', 'Total_counts', 'Total_counts_unclassified', 'Total_counts_classified', 'unique_taxa_per_spot'\n",
      "      uns:\t'log1p'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "    cell2loc:\t604 x 21\n",
      "      obs:\t'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "      obsm:\t'spatial', 'classification'\n",
      "Sample STRS_A:\n",
      "                      A      B     C\n",
      "AAACACCAATAACTGC  False  False  True\n",
      "AAACAGCTTTCAGAAG  False  False  True\n",
      "AAACAGGGTCTATATT  False  False  True\n",
      "AAACATGGTGAGAGGA  False  False  True\n",
      "AAACCGGGTAGGTACC  False   True  True\n",
      "Sample STRS_B:\n",
      "                      A      B     C\n",
      "AAACAAGTATCTCCCA  False  False  True\n",
      "AAACAGTGTTCCTGGG  False  False  True\n",
      "AAACCCGAACGAAATC  False  False  True\n",
      "AAACCGGAAATGTTAA  False  False  True\n",
      "AAACCGTTCGTCCAGG  False  False  True\n",
      "Sample STRS_C:\n",
      "                      A      B     C\n",
      "AAACAAGTATCTCCCA  False   True  True\n",
      "AAACAATCTACTAGCA  False  False  True\n",
      "AAACATTTCCCGGATT  False  False  True\n",
      "AAACGAAGAACATACC  False   True  True\n",
      "AAACGAAGATGGAGTA  False   True  True\n",
      "Sample STRS_D:\n",
      "                      A     B     C\n",
      "AAACAGGGTCTATATT  False  True  True\n",
      "AAACATTTCCCGGATT  False  True  True\n",
      "AAACCGGGTAGGTACC  False  True  True\n",
      "AAACGAAGATGGAGTA  False  True  True\n",
      "AAACGAGACGGTTGAT  False  True  True\n",
      "Sample Visium_A:\n",
      "                      A      B     C\n",
      "AAACAATCTACTAGCA  False  False  True\n",
      "AAACAGAGCGACTCCT  False  False  True\n",
      "AAACAGGGTCTATATT  False  False  True\n",
      "AAACCACTACACAGAT  False  False  True\n",
      "AAACCGGGTAGGTACC  False  False  True\n",
      "Sample Visium_B:\n",
      "                      A      B     C\n",
      "AAACAACGAATAGTTC  False  False  True\n",
      "AAACAATCTACTAGCA  False   True  True\n",
      "AAACACCAATAACTGC  False  False  True\n",
      "AAACAGAGCGACTCCT  False  False  True\n",
      "AAACCACTACACAGAT  False  False  True\n",
      "Sample Visium_C:\n",
      "                      A      B     C\n",
      "AAACACCAATAACTGC  False   True  True\n",
      "AAACCACTACACAGAT  False   True  True\n",
      "AAACCGGAAATGTTAA  False  False  True\n",
      "AAACCGGGTAGGTACC  False   True  True\n",
      "AAACGAAGATGGAGTA  False   True  True\n",
      "Sample Visium_D:\n",
      "                      A      B     C\n",
      "AAACGAGACGGTTGAT  False   True  True\n",
      "AAACTAACGTGGCGAC  False  False  True\n",
      "AAACTGCTGGCTCCAA  False   True  True\n",
      "AAAGGCTACGGACCAT  False  False  True\n",
      "AAAGGCTCTCGCGCCG   True  False  True\n",
      "Sample STRS_A:\n",
      "AnnData object with n_obs × n_vars = 2701 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample STRS_B:\n",
      "AnnData object with n_obs × n_vars = 1662 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample STRS_C:\n",
      "AnnData object with n_obs × n_vars = 1642 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample STRS_D:\n",
      "AnnData object with n_obs × n_vars = 699 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample Visium_A:\n",
      "AnnData object with n_obs × n_vars = 1553 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample Visium_B:\n",
      "AnnData object with n_obs × n_vars = 1725 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample Visium_C:\n",
      "AnnData object with n_obs × n_vars = 1056 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n",
      "Sample Visium_D:\n",
      "AnnData object with n_obs × n_vars = 604 × 21\n",
      "    obs: 'pct.TEC', 'pct.snRNA', 'pct.protein_coding', 'pct.processed_pseudogene', 'pct.lncRNA', 'pct.miRNA', 'pct.snoRNA', 'pct.misc_RNA', 'pct.transcribed_unprocessed_pseudogene', 'pct.unprocessed_pseudogene', 'pct.rRNA', 'pct.transcribed_processed_pseudogene', 'pct.ribozyme', 'pct.unitary_pseudogene', 'pct.scaRNA', 'pct.pseudogene', 'pct.transcribed_unitary_pseudogene', 'pct.translated_unprocessed_pseudogene', 'pct.TR_V_gene', 'pct.TR_V_pseudogene', 'pct.TR_D_gene', 'pct.TR_J_gene', 'pct.TR_C_gene', 'pct.TR_J_pseudogene', 'pct.IG_LV_gene', 'pct.IG_V_gene', 'pct.IG_V_pseudogene', 'pct.IG_J_gene', 'pct.IG_C_gene', 'pct.sRNA', 'pct.scRNA', 'pct.IG_C_pseudogene', 'pct.IG_D_gene', 'pct.IG_D_pseudogene', 'pct.IG_pseudogene', 'pct.Mt_tRNA', 'pct.Mt_rRNA', 'Total_counts', 'Total_counts_spliced', 'Total_counts_unspliced', 'Unspliced_Spliced_ratio', 'Ratio', 'Total_molecules_detected', 'Sample', 'assay', 'n_counts', '_indices', '_scvi_batch', '_scvi_labels', 'total RNA counts', 'mean_nUMI_factorsfact_0', 'mean_nUMI_factorsfact_1', 'mean_nUMI_factorsfact_2', 'mean_nUMI_factorsfact_3', 'mean_nUMI_factorsfact_4', 'mean_nUMI_factorsfact_5', 'mean_nUMI_factorsfact_6', 'mean_nUMI_factorsfact_7', 'mean_nUMI_factorsfact_8', 'mean_nUMI_factorsfact_9', 'mean_nUMI_factorsfact_10', 'imagecol', 'imagerow', 'mean_nUMI_factorsfact_11', 'B cells', 'CD8 T cells', 'Cholangiocytes', 'Enteroendocrine cells', 'Epithelial cells', 'Erythroid-like and erythroid precursor cells', 'Glandular epithelial cells', 'Goblet cells', 'Immature Enterocytes 1', 'Macrophages', 'Mature Enterocytes 1', 'Mature Enterocytes 2', 'Mature Enterocytes 3', 'Mature Enterocytes 4', 'Mature Enterocytes 5', 'Memory T cells', 'Mesothelial cells', 'Paneth cells', 'TA cells', 'Tuft cells', 'pDCs', 'total_abundance', 'max_pred', 'max_pred_celltype'\n",
      "    obsm: 'spatial', 'classification'\n"
     ]
    }
   ],
   "source": [
    "# Also we are gonna add the cell2loc as cluster \n",
    "adata = sc.read_h5ad(\"/fs/cbsuvlaminck3/workdir/in68/STRS-HD/Gut_STRS_060524/CoLocatedComb/CoLocatedGroupsSklearnNMF_11642locations_21factors/anndata/sp.h5ad\")\n",
    "adata.obs[adata.uns['mod']['factor_names']] = adata.obsm['q05_cell_abundance_w_sf']\n",
    "adata.obsm['q05_cell_proportions'] = adata.obsm['q05_cell_abundance_w_sf'].div(adata.obsm['q05_cell_abundance_w_sf'].sum(axis=1), axis=0)\n",
    "\n",
    "# add 5% quantile, representing confident cell abundance, 'at least this amount is present',\n",
    "# to adata.obs with nice names for plotting\n",
    "\n",
    "adata.obs[\"total_abundance\"] = adata.obsm['q05_cell_abundance_w_sf'].sum(axis = 1)\n",
    "# adata_obj.obs[adata_obj.uns['mod']['factor_names']] = adata_obj.obsm['q05_cell_abundance_w_sf']\n",
    "adata.obs[adata.uns['mod']['factor_names']] = adata.obsm['q05_cell_proportions']\n",
    "adata.obs[\"max_pred\"] = adata.obs[adata.uns['mod']['factor_names']].max(axis=1)\n",
    "adata.obs[\"max_pred_celltype\"] = adata.obs[adata.uns['mod']['factor_names']].idxmax(axis=1)\n",
    "adata.obs[\"max_pred_celltype\"] = adata.obs[\"max_pred_celltype\"].astype('category')\n",
    "\n",
    "# Function to revert observation names to original\n",
    "def revert_obs_names(obs_names):\n",
    "    return [name.split('_')[0] for name in obs_names]\n",
    "\n",
    "# Dictionary to store the 'cell2loc' modality for each sample\n",
    "cell2loc_dict = {}\n",
    "\n",
    "# Loop through each sample in the mudata_dict\n",
    "for sample in mudata_dict.keys():\n",
    "    # Extract the data for the current sample from the combined AnnData\n",
    "    sample_data = adata[adata.obs['Sample'] == sample].copy()\n",
    "\n",
    "    # Revert observation names to original for both obs and obsm\n",
    "    reverted_obs_names = revert_obs_names(sample_data.obs.index.tolist())\n",
    "    sample_data.obs.index = reverted_obs_names\n",
    "    sample_data.obsm['q05_cell_abundance_w_sf'].index = reverted_obs_names\n",
    "    sample_data.obsm['spatial'] = pd.DataFrame(sample_data.obsm['spatial'], index=reverted_obs_names)\n",
    "    sample_data.obsm['classification'] = pd.DataFrame(sample_data.obsm['classification'], index=reverted_obs_names)\n",
    "\n",
    "    # Prepare the new AnnData for the 'cell2loc' modality\n",
    "    cell2loc_adata = sc.AnnData(\n",
    "        X=sample_data.obsm['q05_cell_abundance_w_sf'].values,\n",
    "        obs=sample_data.obs,\n",
    "        var=pd.DataFrame(index=sample_data.obsm['q05_cell_abundance_w_sf'].columns)\n",
    "    )\n",
    "\n",
    "    # Ensure the indices are the same order for spatial and classification\n",
    "    spatial_data = sample_data.obsm['spatial'].values\n",
    "    classification_data = sample_data.obsm['classification'].values\n",
    "\n",
    "    # Assign obsm values\n",
    "    cell2loc_adata.obsm['spatial'] = spatial_data\n",
    "    cell2loc_adata.obsm['classification'] = classification_data\n",
    "\n",
    "    # Store the new AnnData object in the dictionary\n",
    "    cell2loc_dict[sample] = cell2loc_adata\n",
    "\n",
    "# Output which samples have been successfully processed\n",
    "for sample, adata in cell2loc_dict.items():\n",
    "    print(f\"Sample {sample}:\")\n",
    "    print(adata)\n",
    "\n",
    "import muon as mu\n",
    "\n",
    "# Loop through each sample in the mudata_dict\n",
    "for sample in mudata_dict.keys():\n",
    "    # Check if the sample exists in the cell2loc_dict\n",
    "    if sample in cell2loc_dict:\n",
    "        # Get the existing MuData object\n",
    "        mudata = mudata_dict[sample]\n",
    "        \n",
    "        # Remove the existing 'cell2loc' modality if it exists\n",
    "        if 'cell2loc' in mudata.mod:\n",
    "            del mudata.mod['cell2loc']\n",
    "        \n",
    "        # Add the new 'cell2loc' modality\n",
    "        mudata.mod['cell2loc'] = cell2loc_dict[sample]\n",
    "        \n",
    "        # Replace the existing MuData object with the updated one\n",
    "        mudata_dict[sample] = mudata\n",
    "\n",
    "# Output the updated MuData objects to confirm the addition\n",
    "for sample, mudata in mudata_dict.items():\n",
    "    print(f\"Sample {sample}:\")\n",
    "    print(mudata)\n",
    "# Loop through each sample in the mudata_dict to add the 'spatial' obsm to 'cell2loc' modality\n",
    "for sample in mudata_dict.keys():\n",
    "    # Get the existing MuData object\n",
    "    mudata = mudata_dict[sample]\n",
    "    \n",
    "    # Ensure 'cell2loc' modality exists in the MuData object\n",
    "    if 'cell2loc' in mudata.mod:\n",
    "        # Get the 'host' and 'cell2loc' AnnData objects\n",
    "        host_adata = mudata['host']\n",
    "        cell2loc_adata = mudata['cell2loc']\n",
    "        \n",
    "        # Ensure the index names match before copying 'spatial' obsm\n",
    "        if set(cell2loc_adata.obs.index).issubset(set(host_adata.obs.index)):\n",
    "            cell2loc_adata.obsm['spatial'] = host_adata.obsm['spatial'][host_adata.obs.index.isin(cell2loc_adata.obs.index), :]\n",
    "        else:\n",
    "            print(f\"Warning: Index mismatch for sample {sample}\")\n",
    "\n",
    "        # Replace the updated 'cell2loc' AnnData object in the MuData object\n",
    "        mudata.mod['cell2loc'] = cell2loc_adata\n",
    "        \n",
    "        # Replace the existing MuData object with the updated one\n",
    "        mudata_dict[sample] = mudata\n",
    "\n",
    "# Output the updated MuData objects to confirm the addition\n",
    "for sample, mudata in mudata_dict.items():\n",
    "    print(f\"Sample {sample}:\")\n",
    "    print(mudata)\n",
    "\n",
    "# Loop through each sample in the mudata_dict\n",
    "for sample in mudata_dict.keys():\n",
    "    # Extract the host and cell2loc data\n",
    "    host_data = mudata_dict[sample]['host']\n",
    "    cell2loc_data = mudata_dict[sample]['cell2loc']\n",
    "    \n",
    "    # Filter the classification data based on the indices in the cell2loc obs\n",
    "    filtered_classification = host_data.obsm['classification'].loc[cell2loc_data.obs.index]\n",
    "    \n",
    "    # Assign the filtered classification data to the cell2loc modality\n",
    "    cell2loc_data.obsm['classification'] = filtered_classification\n",
    "\n",
    "    # Optionally, verify the operation\n",
    "    print(f\"Sample {sample}:\")\n",
    "    print(cell2loc_data.obsm['classification'].head())\n",
    "\n",
    "# Output the updated MuData objects to confirm the addition\n",
    "for sample, mudata in mudata_dict.items():\n",
    "    print(f\"Sample {sample}:\")\n",
    "    print(mudata['cell2loc'])\n",
    "\n",
    "# Original dictionary with colors\n",
    "celltype_color_dict = {\n",
    "    'q05cell_abundance_w_sf_B cells': '#183E9F',\n",
    "    'q05cell_abundance_w_sf_CD8 T cells': '#F978E9',\n",
    "    'q05cell_abundance_w_sf_Cholangiocytes': '#747475',\n",
    "    'q05cell_abundance_w_sf_Enteroendocrine cells': '#F2E64C',\n",
    "    'q05cell_abundance_w_sf_Epithelial cells': '#FC8900',\n",
    "    'q05cell_abundance_w_sf_Erythroid-like and erythroid precursor cells': '#5FCCC0',\n",
    "    'q05cell_abundance_w_sf_Glandular epithelial cells': '#B8F4F2',\n",
    "    'q05cell_abundance_w_sf_Goblet cells': '#8FF469',\n",
    "    'q05cell_abundance_w_sf_Immature Enterocytes 1': '#B6BBE0',\n",
    "    'q05cell_abundance_w_sf_Macrophages': '#BCAE04',\n",
    "    'q05cell_abundance_w_sf_Mature Enterocytes 1': '#C95F7B',\n",
    "    'q05cell_abundance_w_sf_Mature Enterocytes 2': '#C95F7B',\n",
    "    'q05cell_abundance_w_sf_Mature Enterocytes 3': '#C95F7B',\n",
    "    'q05cell_abundance_w_sf_Mature Enterocytes 4': '#C95F7B',\n",
    "    'q05cell_abundance_w_sf_Mature Enterocytes 5': '#C95F7B',\n",
    "    'q05cell_abundance_w_sf_Memory T cells': '#6ABF4B',\n",
    "    'q05cell_abundance_w_sf_Mesothelial cells': '#E8BB93',\n",
    "    'q05cell_abundance_w_sf_Paneth cells': '#6B845E',\n",
    "    'q05cell_abundance_w_sf_TA cells': '#CC0202',\n",
    "    'q05cell_abundance_w_sf_Tuft cells': '#7D199E',\n",
    "    'q05cell_abundance_w_sf_pDCs': '#D9E9E7',\n",
    "    'Others': '#747475'\n",
    "}\n",
    "\n",
    "# Remove prefix\n",
    "prefix = 'q05cell_abundance_w_sf_'\n",
    "celltype_color_dict_no_prefix = {k.replace(prefix, ''): v for k, v in celltype_color_dict.items()}\n",
    "\n",
    "# Function to assign colors to unique cell types in a given AnnData object\n",
    "def assign_colors(adata, celltype_color_dict):\n",
    "    unique_celltypes = np.unique(adata.obs['max_pred_celltype'])\n",
    "    celltype_colors = [celltype_color_dict.get(celltype, celltype_color_dict['Others']) for celltype in unique_celltypes]\n",
    "    adata.uns['max_pred_celltype_colors'] = celltype_colors\n",
    "    return unique_celltypes, celltype_colors\n",
    "\n",
    "# Loop through each sample in the mudata_dict and assign colors\n",
    "for sample in mudata_dict.keys():\n",
    "    adata = mudata_dict[sample]['cell2loc']\n",
    "    unique_celltypes, celltype_colors = assign_colors(adata, celltype_color_dict_no_prefix)\n",
    "    \n",
    "    # # Print the assigned colors for verification\n",
    "    # print(f\"Sample {sample}:\")\n",
    "    # print(\"Unique cell types:\", unique_celltypes)\n",
    "    # print(\"Assigned colors:\", celltype_colors)\n",
    "    \n",
    "    # # Plot to verify\n",
    "    # plt.figure(figsize=(10, 2))\n",
    "    # for i, (celltype, color) in enumerate(zip(unique_celltypes, celltype_colors)):\n",
    "    #     plt.plot([], [], marker='o', markersize=10, label=celltype, color=color)\n",
    "    # plt.legend(loc='center', ncol=4)\n",
    "    # plt.axis('off')\n",
    "    # plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c60d8f2c-8ee0-4185-8207-9992555ee921",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "No gene_version1 genes found...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "Added spatial coordinates for 4887 cells.\n",
      "No data found for bacteria at species level. Skipping this level.\n",
      "No data found for bacteria at genus level. Skipping this level.\n",
      "No data found for bacteria at family level. Skipping this level.\n",
      "No data found for bacteria at phylum level. Skipping this level.\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "No sRNA genes found...\n",
      "No gene_version1 genes found...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Added spatial coordinates for 4992 cells.\n",
      "Processed samples: ['SRR17048630', 'Vis_yPAP_3A_S1_L001']\n"
     ]
    }
   ],
   "source": [
    "## Add the heart controls w and wo polyA\n",
    "# Load the taxonomic identifiers for bacteria from a DataFrame\n",
    "bacteria_taxids = Taxonomic_df[Taxonomic_df['superkingdom'] == 'Bacteria']['tax_id']\n",
    "\n",
    "# Define the base directory where the host and microbial data are stored\n",
    "base_dir = \"/fs/cbsuvlaminck2/workdir/in68/microSTRS/results_out_Downloaded/Controls/\"\n",
    "\n",
    "# List all the sample identifiers\n",
    "samples = ['SRR17048630', 'Vis_yPAP_3A_S1_L001']\n",
    "\n",
    "# Dictionary to store the combined data objects (MuData) for each sample\n",
    "mudata_dict_3 = {}\n",
    "\n",
    "for sample in samples:\n",
    "    try:\n",
    "        # Assemble the directory path for the host data for the current sample\n",
    "        host_data_dir = os.path.join(base_dir, f\"{sample}_solo/Solo.out/GeneFull/raw/\")\n",
    "        spliced_data_dir = os.path.join(base_dir, f\"{sample}_solo/Solo.out/Velocyto/raw/spliced.mtx.gz\")\n",
    "        unspliced_data_dir = os.path.join(base_dir, f\"{sample}_solo/Solo.out/Velocyto/raw/unspliced.mtx.gz\")\n",
    "        \n",
    "        # Check if the host data directory exists; skip the sample if it doesn't\n",
    "        if not os.path.exists(host_data_dir):\n",
    "            print(f\"Host data directory does not exist: {host_data_dir}\")\n",
    "            continue\n",
    "\n",
    "        # Read the host gene expression data using the 10x Genomics format\n",
    "        adata_host = sc.read_10x_mtx(host_data_dir, var_names='gene_symbols')\n",
    "        adata_host = normalize_spatial_coordinates(adata_host, spatial_coord_path)\n",
    "        adata_host.raw = adata_host.copy()\n",
    "        \n",
    "        # Add biotypes percentage and total counts to the host data\n",
    "        add_biotypes_pct(adata_host, gtf_info)\n",
    "        adata_host.obs['Total_counts'] = adata_host.X.sum(axis=1)\n",
    "\n",
    "        # Normalize total gene expression and log transform for host data\n",
    "        sc.pp.normalize_total(adata_host, target_sum=1e4)\n",
    "        sc.pp.log1p(adata_host)\n",
    "\n",
    "        # Read spliced and unspliced data\n",
    "        adata_spliced = sc.read_mtx(spliced_data_dir)\n",
    "        adata_unspliced = sc.read_mtx(unspliced_data_dir)\n",
    "        adata_spliced = adata_spliced.T\n",
    "        adata_unspliced = adata_unspliced.T\n",
    "        adata_spliced.obs = adata_host.obs.copy()\n",
    "        adata_unspliced.obs = adata_host.obs.copy()\n",
    "        adata_spliced.var = adata_host.var.copy()\n",
    "        adata_unspliced.var = adata_host.var.copy()\n",
    "        adata_host.obs['Total_counts_spliced'] = adata_spliced.X.sum(axis=1)\n",
    "        adata_host.obs['Total_counts_unspliced'] = adata_unspliced.X.sum(axis=1)\n",
    "        adata_host.obs['Unspliced.pct'] = adata_host.obs['Total_counts_unspliced']/(adata_host.obs['Total_counts_spliced']+adata_host.obs['Total_counts_unspliced'])\n",
    "\n",
    "        # Construct the path for microbial data and verify it\n",
    "        microbial_data_path = os.path.join(base_dir, f\"{sample}_solo/microbe.h5ad\")\n",
    "        \n",
    "        # Check if the microbial data file exists; skip the sample if it doesn't\n",
    "        if not os.path.exists(microbial_data_path):\n",
    "            print(f\"Microbial data file does not exist: {microbial_data_path}\")\n",
    "            continue\n",
    "\n",
    "        # Read the microbial data\n",
    "        adata_unmapped = sc.read_h5ad(microbial_data_path)\n",
    "        adata_unmapped = normalize_spatial_coordinates(adata_unmapped, spatial_coord_path)\n",
    "\n",
    "        # Store the raw microbial data\n",
    "        adata_unmapped.raw = adata_unmapped.copy()\n",
    "        # adata_unmapped.uns['taxonomy'] = df.copy()\n",
    "        adata_unmapped.obs['Total_counts_unfiltered'] = adata_unmapped.X.sum(axis=1)\n",
    "\n",
    "        # Filter the microbial data to include only bacteria\n",
    "        mask = adata_unmapped.var.index.isin(bacteria_taxids)\n",
    "        adata_bacteria = adata_unmapped[:, mask].copy()\n",
    "        adata_bacteria.obs['Total_counts_unfiltered'] = adata_bacteria.X.sum(axis=1)\n",
    "\n",
    "        # Delete the taxids that are found in controls\n",
    "        mask2 = adata_bacteria.var.index.isin(taxid_to_delete)\n",
    "        adata_bacteria = adata_bacteria[:, ~mask2].copy()\n",
    "\n",
    "\n",
    "        \n",
    "        # Filter the microbial data to include only virus\n",
    "        mask = adata_unmapped.var.index.isin(viruses_taxids)\n",
    "        adata_virus = adata_unmapped[:, mask].copy()\n",
    "        adata_virus.obs['Total_counts_unfiltered'] = adata_virus.X.sum(axis=1)\n",
    "\n",
    "        # Delete the taxids that are found in controls from virus\n",
    "        mask2 = adata_virus.var.index.isin(taxid_to_delete)\n",
    "        adata_virus = adata_virus[:, ~mask2].copy()\n",
    "        adata_virus.obs['Total_counts'] = adata_virus.X.sum(axis=1)\n",
    "\n",
    "\n",
    "        \n",
    "        # Filter the microbial data to include only ARCHAEA\n",
    "        mask = adata_unmapped.var.index.isin(archaea_taxids)\n",
    "        adata_archaea = adata_unmapped[:, mask].copy()\n",
    "        adata_archaea.obs['Total_counts_unfiltered'] = adata_archaea.X.sum(axis=1)\n",
    "\n",
    "        # Delete the taxids that are found in controls from virus\n",
    "        mask2 = adata_archaea.var.index.isin(taxid_to_delete)\n",
    "        adata_archaea = adata_archaea[:, ~mask2].copy()\n",
    "        adata_archaea.obs['Total_counts'] = adata_archaea.X.sum(axis=1)\n",
    "\n",
    "        adata_host.obs['Total_molecules_detected'] = adata_host.obs['Total_counts'] + adata_unmapped.obs['Total_counts_unfiltered']\n",
    "        adatas = {'host': adata_host, 'bacteria': adata_bacteria.copy(),\n",
    "                  'viruses': adata_virus.copy(),'archaea': adata_archaea.copy(), \n",
    "                  'unmapped_all': adata_unmapped.copy(), 'spliced': adata_spliced, 'unspliced': adata_unspliced}\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "        # Process each taxonomic level bacteria\n",
    "        for tax_level in ['species', 'genus', 'family', 'phylum']:\n",
    "            # adata_bacteria.uns['taxonomy'][tax_level] = adata_bacteria.uns['taxonomy'][tax_level].fillna('unclassified')\n",
    "            tax_id_to_name = dict(zip(Taxonomic_df['tax_id'], Taxonomic_df[tax_level]))\n",
    "            adata_bacteria.var[tax_level] = adata_bacteria.var.index.map(tax_id_to_name).fillna('unclassified')\n",
    "            adata_df = pd.DataFrame(adata_bacteria.X.toarray(), columns=adata_bacteria.var[tax_level])\n",
    "            aggregated_data = adata_df.groupby(adata_df.columns, axis=1).sum()\n",
    "            if aggregated_data.shape[1] == 0:\n",
    "                print(f\"No data found for bacteria at {tax_level} level. Skipping this level.\")\n",
    "                continue\n",
    "            new_vars = pd.DataFrame(index=aggregated_data.columns)\n",
    "            X_mat = csr_matrix(np.round(aggregated_data.values).astype(np.float32))\n",
    "\n",
    "            new_adata_bacteria = sc.AnnData(X=X_mat, var=new_vars, obs=adata_bacteria.obs)\n",
    "\n",
    "            # Add the raw component\n",
    "            new_adata_bacteria.raw = new_adata_bacteria.copy()\n",
    "\n",
    "            new_adata_bacteria.obs['Total_counts'] = new_adata_bacteria.X.sum(axis=1)\n",
    "            unclassified_mask = new_adata_bacteria.var.index == 'unclassified'\n",
    "            new_adata_bacteria.obs['Total_counts_unclassified'] = new_adata_bacteria[:, unclassified_mask].X.sum(axis=1)\n",
    "            new_adata_bacteria.obs['Total_counts_classified'] = new_adata_bacteria.obs['Total_counts'] - new_adata_bacteria.obs['Total_counts_unclassified']\n",
    "\n",
    "            # Normalize counts and log transform\n",
    "            total_microbial_counts = new_adata_bacteria.X.sum()\n",
    "            new_adata_bacteria.X = new_adata_bacteria.X / total_microbial_counts * 10000\n",
    "            new_adata_bacteria = normalize_spatial_coordinates(new_adata_bacteria, spatial_coord_path)\n",
    "            spot_counts = new_adata_bacteria.X > 0\n",
    "            unique_genes_per_spot = spot_counts.sum(axis=1)\n",
    "            new_adata_bacteria.obs['unique_taxa_per_spot'] = unique_genes_per_spot\n",
    "            sc.pp.log1p(new_adata_bacteria)\n",
    "\n",
    "            adatas[tax_level] = new_adata_bacteria\n",
    "            # Delete intermediate variables to free up memory\n",
    "            del adata_df, aggregated_data, new_vars, X_mat, new_adata_bacteria, tax_id_to_name            \n",
    "\n",
    "        # Process each taxonomic level for virus\n",
    "        for tax_level in ['species', 'genus', 'family', 'phylum']:\n",
    "            # adata_virus.uns['taxonomy'][tax_level] = adata_virus.uns['taxonomy'][tax_level].fillna('unclassified')\n",
    "            tax_id_to_name = dict(zip(Taxonomic_df['tax_id'], Taxonomic_df[tax_level]))\n",
    "            adata_virus.var[tax_level] = adata_virus.var.index.map(tax_id_to_name).fillna('unclassified')\n",
    "            adata_df = pd.DataFrame(adata_virus.X.toarray(), columns=adata_virus.var[tax_level])\n",
    "            aggregated_data = adata_df.groupby(adata_df.columns, axis=1).sum()\n",
    "            if aggregated_data.shape[1] == 0:\n",
    "                print(f\"No data found for bacteria at {tax_level} level. Skipping this level.\")\n",
    "                continue\n",
    "            new_vars = pd.DataFrame(index=aggregated_data.columns)\n",
    "            X_mat = csr_matrix(np.round(aggregated_data.values).astype(np.float32))\n",
    "\n",
    "            new_adata_virus = sc.AnnData(X=X_mat, var=new_vars, obs=adata_virus.obs)\n",
    "\n",
    "            # Add the raw component\n",
    "            new_adata_virus.raw = new_adata_virus.copy()\n",
    "\n",
    "            new_adata_virus.obs['Total_counts'] = new_adata_virus.X.sum(axis=1)\n",
    "            unclassified_mask = new_adata_virus.var.index == 'unclassified'\n",
    "            new_adata_virus.obs['Total_counts_unclassified'] = new_adata_virus[:, unclassified_mask].X.sum(axis=1)\n",
    "            new_adata_virus.obs['Total_counts_classified'] = new_adata_virus.obs['Total_counts'] - new_adata_virus.obs['Total_counts_unclassified']\n",
    "\n",
    "            # Normalize counts and log transform\n",
    "            total_microbial_counts = new_adata_virus.X.sum()\n",
    "            new_adata_virus.X = new_adata_virus.X / total_microbial_counts * 10000\n",
    "            new_adata_virus = normalize_spatial_coordinates(new_adata_virus, spatial_coord_path)\n",
    "            spot_counts = new_adata_virus.X > 0\n",
    "            unique_genes_per_spot = spot_counts.sum(axis=1)\n",
    "            new_adata_virus.obs['unique_taxa_per_spot'] = unique_genes_per_spot\n",
    "            sc.pp.log1p(new_adata_virus)\n",
    "            tax_level_to_use = f\"{tax_level}_virus\"\n",
    "            adatas[tax_level_to_use] = new_adata_virus\n",
    "            # Delete intermediate variables to free up memory\n",
    "            del adata_df, aggregated_data, new_vars, X_mat, new_adata_virus, tax_id_to_name\n",
    "            \n",
    "        # Process each taxonomic level for archaea\n",
    "        for tax_level in ['species', 'genus', 'family', 'phylum']:\n",
    "            # adata_archaea.uns['taxonomy'][tax_level] = adata_archaea.uns['taxonomy'][tax_level].fillna('unclassified')\n",
    "            tax_id_to_name = dict(zip(Taxonomic_df['tax_id'], Taxonomic_df[tax_level]))\n",
    "            adata_archaea.var[tax_level] = adata_archaea.var.index.map(tax_id_to_name).fillna('unclassified')\n",
    "            adata_df = pd.DataFrame(adata_archaea.X.toarray(), columns=adata_archaea.var[tax_level])\n",
    "            aggregated_data = adata_df.groupby(adata_df.columns, axis=1).sum()\n",
    "            if aggregated_data.shape[1] == 0:\n",
    "                print(f\"No data found for bacteria at {tax_level} level. Skipping this level.\")\n",
    "                continue\n",
    "            new_vars = pd.DataFrame(index=aggregated_data.columns)\n",
    "            X_mat = csr_matrix(np.round(aggregated_data.values).astype(np.float32))\n",
    "\n",
    "            new_adata_archaea = sc.AnnData(X=X_mat, var=new_vars, obs=adata_archaea.obs)\n",
    "\n",
    "            # Add the raw component\n",
    "            new_adata_archaea.raw = new_adata_archaea.copy()\n",
    "\n",
    "            new_adata_archaea.obs['Total_counts'] = new_adata_archaea.X.sum(axis=1)\n",
    "            unclassified_mask = new_adata_archaea.var.index == 'unclassified'\n",
    "            new_adata_archaea.obs['Total_counts_unclassified'] = new_adata_archaea[:, unclassified_mask].X.sum(axis=1)\n",
    "            new_adata_archaea.obs['Total_counts_classified'] = new_adata_archaea.obs['Total_counts'] - new_adata_archaea.obs['Total_counts_unclassified']\n",
    "\n",
    "            # Normalize counts and log transform\n",
    "            total_microbial_counts = new_adata_archaea.X.sum()\n",
    "            new_adata_archaea.X = new_adata_archaea.X / total_microbial_counts * 10000\n",
    "            new_adata_archaea = normalize_spatial_coordinates(new_adata_archaea, spatial_coord_path)\n",
    "            spot_counts = new_adata_archaea.X > 0\n",
    "            unique_genes_per_spot = spot_counts.sum(axis=1)\n",
    "            new_adata_archaea.obs['unique_taxa_per_spot'] = unique_genes_per_spot\n",
    "            sc.pp.log1p(new_adata_archaea)\n",
    "            tax_level_to_use = f\"{tax_level}_archaea\"\n",
    "            adatas[tax_level_to_use] = new_adata_archaea\n",
    "\n",
    "\n",
    "\n",
    "            \n",
    "            # Delete intermediate variables to free up memory\n",
    "            del adata_df, aggregated_data, new_vars, X_mat, new_adata_archaea, tax_id_to_name\n",
    "\n",
    "        # Store the combined data object for each sample\n",
    "        mudata_dict_3[sample] = mu.MuData(adatas)\n",
    "\n",
    "        # Delete intermediate variables to free up memory\n",
    "        del adata_host, adata_unmapped, adata_bacteria, adatas, mask, mask2, adata_spliced, adata_unspliced\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred while processing sample {sample}: {e}\")\n",
    "\n",
    "# Output which samples have been successfully processed\n",
    "print(\"Processed samples:\", list(mudata_dict_3.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "97179a00-4376-45f2-94d0-2b11f70453ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming mudata_dict_2 is already defined and contains your MuData objects\n",
    "\n",
    "# Create a classification DataFrame with 'A' and 'B' as False and 'C' as True\n",
    "def create_classification_df(barcodes):\n",
    "    # Initialize the DataFrame\n",
    "    classification_df = pd.DataFrame(False, index=barcodes, columns=['A', 'B'])\n",
    "    classification_df['C'] = True\n",
    "    return classification_df\n",
    "\n",
    "# Iterate through each MuData object in mudata_dict_3\n",
    "for sample, mudata in mudata_dict_3.items():\n",
    "    for modality in mudata.mod.keys():\n",
    "        adata = mudata[modality]\n",
    "        # Create the classification DataFrame for the barcodes in adata.obs_names\n",
    "        classification_df = create_classification_df(adata.obs_names)\n",
    "        # Add the classification DataFrame to the `obsm` slot\n",
    "        adata.obsm['classification'] = classification_df\n",
    "\n",
    "# mudata_dict_3['SRR17048630']['host'].obsm['classification']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ef82aa13-d2ee-43e2-9a07-f936e8ce1226",
   "metadata": {},
   "outputs": [],
   "source": [
    "# and add them to mudata_dict\n",
    "mudata_dict['Visium_heart'] = mudata_dict_3.pop('SRR17048630')\n",
    "\n",
    "# and add them to mudata_dict\n",
    "mudata_dict['STRS_heart'] = mudata_dict_3.pop('Vis_yPAP_3A_S1_L001')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "13aa3cad-5a04-44bb-ab4f-738b6b6465b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13982"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3c926be",
   "metadata": {},
   "source": [
    "## save adata objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a28bf7ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = \"/workdir/in68/microSTRS/saved_mudata\"\n",
    "os.makedirs(save_dir, exist_ok=True)\n",
    "\n",
    "for sample, mdata in mudata_dict.items():\n",
    "    sample_dir = os.path.join(save_dir, sample)\n",
    "    os.makedirs(sample_dir, exist_ok=True)\n",
    "    for mod_name, ad in mdata.mod.items():\n",
    "        ad.write(os.path.join(sample_dir, f\"{mod_name}.h5ad\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cell2loc_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
