# **microSTRS:** ***Spatial RNA-Sequencing of Microbiome-Host Interactions at One Micron Resolution***
In situ polyadenylation combined with Sequencing based Spatial Transcriptomics allows the mapping of Microbiome + Total Host Transcriptome
![Microbiome_mapping Image](Figures/figure.png)

## Dependencies

The following software and packages were used to run this project:

- **Python** - Version 3.9.17
- **Snakemake** - Version 6.1.0
- **STAR** - Version 2.7.10a
- **Kraken2** - Version 2.09
- **Pandas** - Version 2.1.3
- **Scanpy** - Version 1.9.3
- **NumPy** - Version 1.26.2
- **SciPy** - Version 1.11.3
- **MuData** 
- **Muon** 
- **cell2location** - Version 0.1.3
- **ST_BarcodeMap** - (linked to GitHub; no specific version required)
- **anndata** - Version 0.10.3


### Alignment
- An earlier version (as found in the relevant subdirectory) for the slidesnake pipeline was used for alignment 
### Data generated for this paper:
Data will be made available upon publication under GEO accession numbers; GSE276866 for the low-resolution datasets, GSE277196, GSE277197 and GSE277255 for the high-resolution datasets.


## Contact:
- Iwijn De Vlaminck (id93@cornell.edu)
- Yannis Ntekas (in68@cornell.edu)
- Lena Takayasu (lt425@cornell.edu)
- David W. McKellar (dmckellar@nygenome.org)

The pipeline is in active developement if you have issues/suggestions please contact us! 
